src/main/java/dev/codesupport/web/common/security/JJwtUtility.java

```diff
@@ -0,0 +1,121 @@
+    public JJwtUtility(JwtConfiguration jwtConfiguration) {
+        super(jwtConfiguration);
+    }
+
+    /**
+     * Generates a new token string.
+     * <p>Using the given username, and pre-configured expiration, this method creates a new token string.</p>
+     *
+     * @param username The username to use in the token
+     * @return The encoded token string
+     * @see JwtConfiguration
+     */
+    @Override
+    public String generateToken(String username) {
+        return Jwts.builder()
+                .setIssuer(configuration().getIssuer())
+                .claim(USERNAME, username)
+                .claim(EXPIRATION, calculateExpiration(configuration().getExpiration()))
+                .signWith(KEY)
+                .compact();
+    }
+
+    /**
+     * Decodes a token string to a {@link dev.codesupport.web.common.security.JwtUtility.DecodedJWT}
+     * <p>Validates and decodes a token string.
+     * Throws an exception if the token is a bad format, expired, or otherwise invalid.</p>
+     *
+     * @param tokenString The token string to decode.
+     * @return The {@link dev.codesupport.web.common.security.JwtUtility.DecodedJWT}
+     * @throws InvalidTokenException Thrown if any token validations fail.
+     * @see dev.codesupport.web.common.security.JwtUtility.DecodedJWT
+     * @see DecodedJJWT
+     */
+    @Override
+    DecodedJWT decode(String tokenString) {
+        DecodedJWT decodedJWT;
+        try {
+            decodedJWT = new DecodedJJWT(
+                    Jwts.parser()
+                            .setSigningKey(KEY)
+                            .parseClaimsJws(tokenString)
+            );
+
+            if (decodedJWT.getExpiration() != null &&
+                    decodedJWT.getExpiration() <= System.currentTimeMillis()) {
+                throw new InvalidTokenException(InvalidTokenException.Reason.EXPIRED);
+            }
+            // Catch all expected validation exceptions, and rethrow InvalidTokenException for simplicity.
+        } catch (ExpiredJwtException |
+                UnsupportedJwtException |
+                MalformedJwtException |
+                SignatureException |
+                IllegalArgumentException e) {
+            throw new InvalidTokenException(InvalidTokenException.Reason.INVALID, e);
+        }
+
+        return decodedJWT;
+    }
+
+    /**
+     * {@link JJwtUtility} specific implementation for {@link dev.codesupport.web.common.security.JwtUtility.DecodedJWT}
+     * <p>Handles the io.jsonwebtoken library specific logic for the
+     * {@link dev.codesupport.web.common.security.JwtUtility.DecodedJWT} api methods</p>
+     */
+    private static class DecodedJJWT implements DecodedJWT {
+
+        private final Jws<Claims> jwsClaims;
+
+        DecodedJJWT(Jws<Claims> jwsClaims) {
+            this.jwsClaims = jwsClaims;
+        }
+
+        @Override
+        public String getUsername() {
+            return jwsClaims.getBody().get(USERNAME, String.class);
```

Comment:
We may have other use cases where we need something from the header, thought not to arbitrarily limit the information you can get, but this can be changed if it's desired.

Reply:
Seems like there'd only be benefits from caching it
```java
return this.jwsClaimsBody.get(USERNAME, String.class);
```
Or am I missing something?

Reply:
The only limitation is if we need something from something other then the body, which i'm not sure if we do or not yet.  If you need something from the head it's jwsClaims.getHeader() or signature would be jwsClaims.getSignature()

Reply:
What's the issue with also caching them if need be?

Reply:
It doesn't affect any tests so this should be fine, I've updated the code to cache the body, as it's the only thing we're using now anyways, easy to add more caches for the others if we need them later